When converting HTML that contains an image to an attributed string and back using `HTMLConverter::editingAttributedString`, the image is lost

rr-codes opened the pull request:
#### 71e39c7bd8fc08ef9171cf4bfbaed96673158980
<pre>
When converting HTML that contains an image to an attributed string and back using `HTMLConverter::editingAttributedString`, the image is lost
<a href="https://bugs.webkit.org/show_bug.cgi?id=274701">https://bugs.webkit.org/show_bug.cgi?id=274701</a>
<a href="https://rdar.apple.com/128717614">rdar://128717614</a>

Reviewed by Aditya Keerthi.

When converting an NSAttributedString to an HTML document fragment, if the attributed string
contains a text attachment backed by a file wrapper without a preferred file name, the attachment
is dropped and not converted to HTML.

Fix by ensuring that a file wrapper for an image always has a preferred file name when converting
HTML to an attributed string inside of `HTMLConverter::editingAttributedString` using the following
mechanism:

1. If the image is backed by an attachment, use the attachment&apos;s title as the preferred name.
2. Otherwise, try to use the image&apos;s alt attribute&apos;s value as the preferred name, if present.
3. Otherwise, try to convert the image&apos;s url into a filename and use that as the preferred name.
4. If all else fails, use the default &quot;unknown&quot; file name as the preferred name.

* Source/WebCore/editing/cocoa/HTMLConverter.mm:
(HTMLConverter::HTMLConverter):
(HTMLConverter::_addLinkForElement):
(HTMLConverter::_processElement):
(preferredFilenameForElement):
(fileWrapperForElement):

Canonical link: <a href="https://commits.webkit.org/279425@main">https://commits.webkit.org/279425@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4f3cb8ddff1290e3af0d7c2c55a9d7042ebe5f16

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/53473 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/32829 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/5979 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/56753 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/4199 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/55779 "Passed tests") | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/49/builds/40292 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/3979 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/43341 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/2747 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/55571 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/31013 "Unexpected infrastructure issue, retrying build") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/46206 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/24476 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/27872 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/3526 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/2355 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/49631 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/3700 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/58348 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/28630 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/3696 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/50743 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/29837 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/46402 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/50087 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [❌ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/11650 "Found 1 new test failure: imported/w3c/web-platform-tests/webrtc/simulcast/getStats.https.html (failure)") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/30765 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/7867 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/29607 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebCore/editing/cocoa/HTMLConverter.mm:
```diff
@@ -2338,11 +2337,39 @@ static NSInteger _colCompare(id block1, id block2, void *)
+    if (!url)
+        url = [NSURL _web_URLWithString:[urlString stringByTrimmingCharactersInSet:[NSCharacterSet whitespaceAndNewlineCharacterSet]] relativeToURL:nil];
+
+    auto* frame = element.document().frame();
```
`RefPtr`

pxlcoder commented on Source/WebCore/editing/cocoa/HTMLConverter.mm:
```diff
@@ -2338,11 +2337,39 @@ static NSInteger _colCompare(id block1, id block2, void *)
+
+    auto* frame = element.document().frame();
+    if (frame->loader().frameHasLoaded()) {
+        DocumentLoader *dataSource = frame->loader().documentLoader();
```
`RefPtr`

pxlcoder reviewed (commented):
> When converting an NSAttributedString to an HTML document fragment, if the attributed string contains a text attachment backed by a file wrapper without a preferred file name, the attachment is dropped and not converted to HTML

Is the "dropping" due to WebKit code, or is it underneath us?

pxlcoder commented on Source/WebCore/editing/cocoa/HTMLConverter.mm:
```diff
@@ -2338,11 +2337,39 @@ static NSInteger _colCompare(id block1, id block2, void *)
 #endif
 
+
+static String preferredFilenameForElement(const HTMLImageElement& element)
```
What's the output for an `<img>` that is not backed by an `<attachment>`, and has a blob URL?

rr-codes replied:
I'm not sure; the non-attachment logic is loosely based on how `attributedString` decides on a file name. How do I make a test page that has an image with a blob url?

pxlcoder replied:
Easiest way is probably to just copy an image from Notes (or another system application) into a `contenteditable` in Safari / Minibrowser.

Otherwise, create a Blob using JS, `URL.createObjectURL`, and assign to an `<img>`'s src.

pxlcoder commented on Source/WebCore/editing/cocoa/HTMLConverter.mm:
```diff
@@ -2338,11 +2337,47 @@ static NSInteger _colCompare(id block1, id block2, void *)
+    if (RefPtr attachmentElement = element.attachmentElement())
+        return attachmentElement->attachmentTitle();
+
+    const auto& altText = element.altText();
```
Probably best to just copy this.

pxlcoder commented on Source/WebCore/editing/cocoa/HTMLConverter.mm:
```diff
@@ -2338,11 +2337,47 @@ static NSInteger _colCompare(id block1, id block2, void *)
     if (CachedImage* cachedImage = element.cachedImage()) {
-        if (FragmentedSharedBuffer* sharedBuffer = cachedImage->resourceBuffer())
-            return adoptNS([[NSFileWrapper alloc] initRegularFileWithContents:sharedBuffer->makeContiguous()->createNSData().get()]);
+        if (FragmentedSharedBuffer* sharedBuffer = cachedImage->resourceBuffer()) {
```
```suggestion
        if (RefPtr sharedBuffer = cachedImage->resourceBuffer()) {
```

rr-codes committed:
When converting HTML that contains an image to an attributed string and back using `HTMLConverter::editingAttributedString`, the image is lost
https://bugs.webkit.org/show_bug.cgi?id=274701
rdar://128717614

Reviewed by Aditya Keerthi.

When converting an NSAttributedString to an HTML document fragment, if the attributed string
contains a text attachment backed by a file wrapper without a preferred file name, the attachment
is dropped and not converted to HTML.

Fix by ensuring that a file wrapper for an image always has a preferred file name when converting
HTML to an attributed string inside of `HTMLConverter::editingAttributedString` using the following
mechanism:

1. If the image is backed by an attachment, use the attachment's title as the preferred name.
2. Otherwise, try to use the image's alt attribute's value as the preferred name, if present.
3. Otherwise, try to convert the image's url into a filename and use that as the preferred name.
4. If all else fails, use the default "unknown" file name as the preferred name.

* Source/WebCore/editing/cocoa/HTMLConverter.mm:
(HTMLConverter::HTMLConverter):
(HTMLConverter::_addLinkForElement):
(HTMLConverter::_processElement):
(preferredFilenameForElement):
(fileWrapperForElement):

Canonical link: https://commits.webkit.org/279425@main
